internal/util/cryptsetup.go

```diff
@@ -18,22 +18,41 @@ package util
 
 import (
 	"bytes"
+	"context"
+	"errors"
 	"fmt"
 	"os"
 	"os/exec"
 	"strconv"
 	"strings"
+	"time"
 
 	"github.com/ceph/ceph-csi/internal/util/file"
 	"github.com/ceph/ceph-csi/internal/util/log"
 )
 
-// Limit memory used by Argon2i PBKDF to 32 MiB.
-const cryptsetupPBKDFMemoryLimit = 32 << 10 // 32768 KiB
+const (
+	// Maximum time to wait for cryptsetup commands to complete.
+	CryptSetupExecutionTimeout = 2*time.Minute + 30*time.Second
+
+	// Limit memory used by Argon2i PBKDF to 32 MiB.
+	cryptsetupPBKDFMemoryLimit = 32 << 10 // 32768 KiB
+)
+
+// LuksWrapper is a struct that provides a context-aware wrapper around cryptsetup commands.
+type LuksWrapper struct {
+	ctx context.Context
+}
+
+// NewLuksWrapperWithContext creates a new LuksWrapper instance with the provided context.
+// The context is used to control the lifetime of the cryptsetup commands executed by the LuksWrapper.
+func NewLuksWrapperWithContext(ctx context.Context) *LuksWrapper {
```

Comment:
Same here, `NewLUKSWrapper` would be ok. Also no need to add the `WithContext` as there is only one `New..` call.

To prevent issues with incorrect use, you could do the  following as well:
```go
type LUKSWrapper interface {
    Format(devicePath, passphrase string) (string, string, error)
    Open(...
    ...
}

type luksWrapper struct {
    ctx context.Context
}

// verify that luksWrapper implements the LUKSWrapper interface
var _ LUKSWrapper = &luksWrapper{}

func NewLuksWrapper(ctx context.Context) LuksWrapper {
    return &luksWrapper{ctx: ctx}
}
```

This way, nbody outside the `internal/util` package can create stucts of `luksWrapper` without setting a context.

As this LUKS API is much cleaner (already without the above suggestion), consider moving it all to a `internal/util/cryptsetup` package. Currently `internal/util` has too many things. Moving can be done in a separate PR later on too.

Reply:
I have updated the PR with suggested changes. Should I move `StripSecrets` inside `util/log`? Makes more sense there (plus, we will have one less package).

I skipped `var _ LUKSWrapper = &luksWrapper{}` as `NewLUKSWrapper` is expected to return a type that implements `LUKSWrapper` so the interface verification would be done here. Correct?

Reply:
Yes, that is fine. The error message might be a little more confusing in case there is a mismatch between the interface/struct, but nothing blocking this imho.

Great cleanup, looks much tidier this way.